File: ruby/longest_palindrome.rb
# Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
#
# Example 1:
#
# Input: "babad"
# Output: "bab"
# Note: "aba" is also a valid answer.
# Example 2:
#
# Input: "cbbd"
# Output: "bb"
#
#
# @param {String} s
# @return {String}
#
def longest_palindrome(s)
  start_idx = 0
  end_idx = 0
  i = 0
  while i < s.length
    len1 = expand_around_center(s, i, i)
    len2 = expand_around_center(s, i, i + 1)
    len = [len1, len2].max
    if len > (end_idx - start_idx)
      start_idx = i - (len - 1) / 2
      end_idx = i + len / 2
    end
    i += 1
  end
  s[start_idx..end_idx]
end

def expand_around_center(s, left, right)
  l_idx = left
  r_idx = right

  while l_idx >= 0 && r_idx < s.length && s[l_idx] == s[r_idx]
    l_idx -= 1
    r_idx += 1
  end

  return r_idx - l_idx - 1
end

File: javaScript/longestPalindrome.js
// Given a string s, find the longest palindromic substring in s. You may assume that the maximum length of s is 1000.
//
// Example 1:
//
// Input: "babad"
// Output: "bab"
// Note: "aba" is also a valid answer.
// Example 2:
//
// Input: "cbbd"
// Output: "bb"

/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
  let startIdx = 0;
  let endIdx = 0;

  for (let i = 0; i < s.length; i++) {
    let len1 = expandAroundCenter(s, i, i);
    let len2 = expandAroundCenter(s, i, i + 1);
    let len = Math.max(len1, len2);
    if (len > endIdx - startIdx) {
      startIdx = i - (len - 2) / 2;
      endIdx = i + len / 2;
    }
  }
  return s.slice(startIdx, endIdx + 1);
};

function expandAroundCenter(s, left, right) {
  let leftIdx = left;
  let rightIdx = right;

  while (leftIdx >= 0 && rightIdx < s.length && s[leftIdx] === s[rightIdx]) {
    leftIdx--;
    rightIdx++;
  }

  return rightIdx - leftIdx - 1;
}
